src/pyk/kcfg/kcfg.py

```diff
@@ -854,6 +861,153 @@ def split_on_constraints(self, source_id: NodeIdLike, constraints: Iterable[KInn
         self.create_split(source.id, zip(branch_node_ids, csubsts, strict=True))
         return branch_node_ids
 
+    def lift_edge(self, id: NodeIdLike) -> None:
+        """Lift an edge up another edge directly preceding it.
+
+        Input:
+            -   id: the identifier of the central node `B` of a sequence of edges `A --> B --> C`.
+
+        Effect: `A --M steps--> B --N steps--> C` becomes `A --(M + N) steps--> C`. Node `B` is removed.
+
+        Output: None
+
+        Raises: An `AssertionError` if the edges in question are not in place.
+        """
+
+        # Obtain edges `A -> B`, `B -> C`
+        a_to_b = single(self.edges(target_id=id))
+        b_to_c = single(self.edges(source_id=id))
+
+        # Remove the node `B`, effectively removing the entire initial structure
+        self.remove_node(id)
+
+        # Create edge `A -> C`
+        self.create_edge(a_to_b.source.id, b_to_c.target.id, a_to_b.depth + b_to_c.depth, a_to_b.rules + b_to_c.rules)
+
+    def lift_edges(self) -> bool:
+        """Perform all possible edge lifts across the KCFG.
+
+        Given the KCFG design, it is not possible for one edge lift to either disallow another or
+        allow another that was not previously possible. Therefore, this function is guaranteed to
+        lift all possible edges without having to loop.
+
+        Input: None
+
+        Effect: The KCFG is transformed to an equivalent in which no further edge lifts are possible.
+
+        Output:
+            -   bool: An indicator of whether or not at least one edge lift was performed.
+        """
+
+        edges_to_lift = [
+            node.id
+            for node in self.nodes
+            if len(self.edges(source_id=node.id)) > 0 and len(self.edges(target_id=node.id)) > 0
+        ]
+
+        for node_id in edges_to_lift:
+            self.lift_edge(node_id)
+
+        # The lifting was successful
+        return len(edges_to_lift) > 0
+
+    def lift_split(self, id: NodeIdLike) -> None:
+        """Lift a split up an edge directly preceding it.
+
+        Input:
+            -   id: the identifier of the central node `B` of the structure `A --> B --> [C_1, ..., C_N]`.
+
+        Effect:
+            `A --M steps--> B --[cond_1, ..., cond_N]--> [C_1, ..., C_N]` becomes
+            `A --[cond_1, ..., cond_N]--> [A #And cond_1 --M steps--> C_1, ..., A #And cond_N --M steps--> C_N]`.
+            Node `B` is removed.
+
+        Output: None
+
+        Raises:
+            -   `AssertionError`, if the structure in question is not in place.
+            -   `AssertionError`, if any of the `cond_i` contain variables not present in `A`.
+        """
+
+        # Obtain edge `A -> B`, split `[cond_I, C_I | I = 1..N ]`
+        a_to_b = single(self.edges(target_id=id))
+        a = a_to_b.source
+        split_from_b = single(self.splits(source_id=id))
+        ci, substs = list(split_from_b.splits.keys()), list(split_from_b.splits.values())
```

Comment:
```suggestion
        ci, csubsts = list(split_from_b.splits.keys()), list(split_from_b.splits.values())
```

These are `CSubst`, not `Subst`, so let's name them as such. https://github.com/runtimeverification/pyk/blob/af2567586fcc13163cb8aea695c807deffb1ebb7/src/pyk/cterm/cterm.py#L300